### PR TITLE
FIX label vertical alignment can now be specified

### PR DESCRIPTION
**Pull request by NelleV**
This patch adds a new rcParams allowing to set label vertical alignment. The
sole reason for the existance of this new parameter is to allow user to reset
the style to before 2.0 for testing purposes.

This patch provides a fix to a very pick backward compatibility break. Unfortunately, the
patch that introduces this break also regenerated the test images. This leads to the introduction of a private _test_class stylesheet, corresponding to our current test suite, and a classic stylesheet which is backward compatible.

refs #7905

**TODO**

- [x] Document this new parameter
- [x] add the x equivalent for consistency.

**Note**: this pull request supersedes #7970

### CODE REVIEW

File: lib/matplotlib/rcsetup.py [hunk 1]
```diff
@@ -465,6 +465,11 @@ def validate_font_properties(s):
     'verbose',
     ['silent', 'helpful', 'debug', 'debug-annoying'])
 
+validate_alignment = ValidateInStrings(
```
Comment (NelleV) [inline lib/matplotlib/rcsetup.py]: note to myself. Make this private.

**Comment by NelleV** [comment]
@tacaswell I am having second thoughts about the API here. There are two possible alignments (horizontal and vertical). Should I rename the rcParams to help identify which of the two alignments we are talking about?

File: lib/matplotlib/tests/test_axes.py [hunk 2]
```diff
@@ -1143,6 +1143,15 @@ def test_bar_tick_label_multiple():
     ax.bar([1, 2.5], [1, 2], width=[0.2, 0.5], tick_label=['a', 'b'],
            align='center')
 
+@image_comparison(baseline_images=['bar_tick_label_multiple_old_alignment'],
```
Comment (afvincent) [inline lib/matplotlib/tests/test_axes.py]: `bar_tick_label_multiple_old_label_alignment` (cf. the baseline image name)?

File: lib/matplotlib/tests/test_axes.py [hunk 3]
```diff
@@ -1143,6 +1143,15 @@ def test_bar_tick_label_multiple():
     ax.bar([1, 2.5], [1, 2], width=[0.2, 0.5], tick_label=['a', 'b'],
            align='center')
 
+@image_comparison(baseline_images=['bar_tick_label_multiple_old_label_alignment'],
+                  extensions=['png'])
+def test_bar_tick_label_multiple_old_alignment():
+    # From 2516: plot bar with array of string labels for x axis
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_axes.py]: This test doesn't really have anything to do with #2516.

File: lib/matplotlib/tests/test_backend_pgf.py [hunk 4]
```diff
@@ -78,7 +78,7 @@ def create_figure():
 
 
 # test compiling a figure to pdf with xelatex
-@cleanup(style='classic')
+@cleanup(style='_classic_test')
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_backend_pgf.py]: The tests in this file don't actually use `classic`, as noted in the pytest conversion; maybe just set them to `default` now to reflect what's really going on?

**Comment by NelleV** [comment]
@QuLogic Comments are fixed.

**Comment by tacaswell** [comment]
I see the case for making it 'ytick.valignment' and 'xtick.halignment', however it adds a bit more verbosity, I don't see the use-case for tweaking the other direction (for one thing I think it would break all of the layout logic), and it breaks a nice symmetry between xtick/ytick.

I am :+1: for this going in as-is, and would not protest if the names were changed to be more explicit about what direction they were affecting.

**Pull request merged**